BxlAsLib.txt

```diff
@@ -0,0 +1,377 @@
+    - Some renderer that refers to FileContentInfo of BuildXL.Storage needs to use plain string.
+    - Perhaps create a restricted version of PipDataBuilder that user can use easily.
+  - To support reuse, some classes/enums in BuildXL.Pips needs to be moved to a smaller DLL (may be called
+    BuildXL.PipUtilities), e.g., ProcessOptions, EnvironmentVariables, etc.
+ 
+- ProcessData, WriteFileData, ... are the ones that are going to be serialized/deserialized, and thus need to be versioned.
+
+  - Option 1: Use Google protobuf to represent them.
+    - Pros:
+      - Kind of "de-facto" standard.
+      - Serialization and deserialization come out of box.
+      - Binary format is optimized.
+    - Cons:
+      - Cannot use custom serialization:
+        - We want the serialization to leverage pip inlining mechanism.
+      - Generate only classes.
+      - Deserialization can create a lot of objects.
+
+  - Option 2: Use protobuf-net.
+    - Pros:
+      - Serialization and deserialization come out of box.
+      - Doesn't need a separate proto file.
+    - Cons:
+      - Similar to Google protobuf.
+      - Stability, performance, and encoding size need to be investigated.
+
+  - Option 3 (potentially chosen): Classes have serialization & deserialization methods that mimic
+    protobuf, i.e., there are positional fields.
+    Ex.
+
+    class Data 
+    {       
+        /* TBD: We may want to add positional attrbute for each property, like protobuf-net. */
+
+        public AbsolutePath Path { get; set; }
+        public FileArtifact File { get; set; }
+        public void Serialize(Writer writer) 
+        {
+            if (Path.IsValid) { writer.WritePos(1); writer.WritePath(Path); }
+            if (File.IsValid) { writer.WritePos(2); write.WriteFile(File); }
+        }
+        public static Data Deserialize(Reader reader)
+        {
+            Data data;
+            while (!int pos = reader.ReadPos()) != -1) {
+                switch (pos) {
+                    case 1: { data.Path = reader.ReadPath(); break; }
+                    case 2: { data.File = reader.ReadFile(); break; }
+                }
+            }
+            return data;
+        }
+    }
+
+    - Pros: Can use custom serialization/deserialization.
+    - Cons: 
+      - Manually maintained the position.
+      - An intern working on XLG++ may choose protobuf for versioning.
+
+- Excerpt of ProcessData: basically it reflects Process in BuildXL.Pips
+
+  class ProcessData 
+  {
+      /* TBD: We may want to add positional attrbute for each property, like protobuf-net. */
+
+      public FileArtifact Executable { get; set; }
+      public PipData Arguments { get; set; }
+      public AbsolutePath WorkingDirectory { get; set; }
+      public IEnumerable<EnvironmentVariable> EnvironmentVariables { get; set; }
+      public ProcessOptions ProcessOptions { get; set; }
+      public IEnumerable<FileArtifact> Dependencies { get; set; }

```

Comment:
>IEnumerable [](start = 13, length = 11)

IEnumerable does not allow on demand construction, i.e. composing the values. So either make this a getter only and have an 'AddDependency' method or make it ICollection